### PR TITLE
 bug: Handle 0 messages from data store return

### LINKED ISSUE #8: Panic when reading from an empty store
I've got pushbox up and running successfully and integrated with fxa-auth-server, but if a device trys to read its queue in pushbox before anything as been written to it, then the pushbox server panics with the following traceback:

```
GET /v1/store/2f4baff83fa5491189f39cdc6845aabb/cb0b23567e82b487e84b7a12b2f4ea55?limit=100:
    => Matched: GET /v1/store/<user_id>/<device_id>?<options>
May 29 08:29:52.024 INFO Checking auth token
May 29 08:29:52.026 INFO Handling Read
May 29 08:29:52.026 INFO Found messages, len: 0
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', libcore/option.rs:335:21
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:205
   3: std::panicking::default_hook
             at libstd/panicking.rs:221
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:457
   5: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:344
   6: rust_begin_unwind
             at libstd/panicking.rs:322
   7: core::panicking::panic_fmt
             at libcore/panicking.rs:71
   8: core::panicking::panic
             at libcore/panicking.rs:51
   9: rustbox::server::read_opt
  10: rustbox::server::rocket_route_fn_read_opt
  11: rocket::rocket::Rocket::dispatch
  12: <rocket::rocket::Rocket as hyper::server::Handler>::handle
  13: <hyper::server::Worker<H>>::handle_connection
  14: hyper::server::listener::spawn_with::{{closure}}
```

### PR DESCRIPTION
**Pull request by jrconlin**
Closes #8

### CODE REVIEW

File: src/server.rs [hunk 1]
```diff
@@ -248,8 +248,12 @@ fn read_opt(
         None => true,
         Some(0) => true,
         Some(_) => {
-            let last = messages.last().unwrap();
-            (last.idx as u64) == max_index
+            if messages.len() == 0 {
```
Comment (bbangert) [inline src/server.rs]: ```if messages.is_empty()```

File: src/server.rs [hunk 2]
```diff
@@ -248,8 +248,12 @@ fn read_opt(
...
+                true
+            } else {
+                let last = messages.last().unwrap();
+                (last.idx as u64) == max_index
+            }
```
Comment (bbangert) [inline src/server.rs]: This whole block should probably be:
```rust
messages.last().map(|last| (last.idx as u64) == max_index).unwrap_or(true)
```

Comment (jrconlin) [reply src/server.rs]: Thanks! I keep forgetting that last() is effectively an iterator of one item.

File: src/db/models.rs [hunk 3]
```diff
@@ -80,7 +80,7 @@ impl DatabaseManager {
                 pushboxv1::user_id.eq(user_id),
                 pushboxv1::device_id.eq(device_id),
                 pushboxv1::ttl.eq(ttl as i64),
-                pushboxv1::data.eq(data.clone().into_bytes()),
+                pushboxv1::data.eq(String::from(data).as_bytes()),
```
Comment (pjenvey) [inline src/db/models.rs]: I think just data.as_bytes() will do it?

Comment (jrconlin) [reply src/db/models.rs]: I tried that and rust was not happy, that's why I had to burn a `String`.

File: src/server.rs (hunk 2 above)
Comment (bbangert) [reply src/server.rs]: I meant the entire if/else block, the test for is_empty isn't needed.

**Comment by jrconlin** [comment]
Ah, ok, that definitely makes more sense.

File: src/server.rs [hunk 4]
```diff
@@ -392,7 +422,10 @@ mod test {
     }
 
     fn device_id() -> String {
-        thread_rng().gen_ascii_chars().take(8).collect()
+        iter::repeat(())
```
Comment (pjenvey) [inline src/server.rs]: that sample_iter I mentioned in the last PR's: `thread_rng().sample_iter(&distributions::Alphanumeric).take(8).collect()`

**Review by pjenvey — DISMISSED** [review]

**Review by pjenvey — DISMISSED** [review]

File: src/server.rs [hunk 5]
```diff
@@ -244,14 +244,10 @@ fn read_opt(
...
-        Some(0) => true,
-        Some(_) => {
-            let last = messages.last().unwrap();
-            (last.idx as u64) == max_index
-        }
-    };
+    let is_last = messages
+        .last()
+        .map(|last| (last.idx as u64) == max_index)
+        .unwrap_or(true);
```
Comment (bbangert) [inline src/server.rs]: This doesn't do the same thing as the above block it replaces, if it's possible that messages could be retrieved and the limit is set to None.

**Review by bbangert — APPROVED** [review]

**Pull request merged**